### PR TITLE
Fix #363; Compare Bidirectional Recurrence vs Unidirectional Recurrence

### LINKED ISSUE #363: Compare Bidirectional Recurrence vs Unidirectional Recurrence

### PR DESCRIPTION
**Pull request by andi4191**
@kdavis-mozilla, @reuben : Can you have a look at this?

Changes Made: Replaced the lstm_cell consisting of forward and backward cell with the lstm_cell_stack consisting of 2 forward cells. The dimensionality of single 4096 units from the LSTM cell had been changed to 2 layer of 2048 units.

Dataset used: TED
Trainin time: 20:13:00
No. of Train epochs: 10
GPU instances used: 4 

Results:

Unidirectional Recurrence Case:

------------------------------------------------------------------------
I Test of Epoch 10 - WER: 0.448848, loss: 77.5554098553, mean edit distance: 0.209433
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 1.252548, mean edit distance: 0.076923
I  - src: "no im serious"
I  - res: "now im serious"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 1.650417, mean edit distance: 0.076923
I  - src: "no im serious"
I  - res: "now im serious"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 1.906622, mean edit distance: 0.076923
I  - src: "no im serious"
I  - res: "now im serious"
I --------------------------------------------------------------------------------
I WER: 0.333333, loss: 3.964957, mean edit distance: 0.222222
I  - src: "and i say"
I  - res: "and i save"
I --------------------------------------------------------------------------------
I WER: 0.500000, loss: 2.543439, mean edit distance: 0.100000
I  - src: "farms that"
I  - res: "forms that"
I --------------------------------------------------------------------------------
I WER: 0.500000, loss: 3.561996, mean edit distance: 0.100000
I  - src: "farms that"
I  - res: "forms that"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 1.825124, mean edit distance: 0.333333
I  - src: "and"
I  - res: "nd"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.459296, mean edit distance: 0.500000
I  - src: "so"
I  - res: "o"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 2.511882, mean edit distance: 0.500000
I  - src: "so"
I  - res: "o"
I --------------------------------------------------------------------------------
I WER: 1.000000, loss: 3.144284, mean edit distance: 1.000000
I  - src: "so"
I  - res: ""
I --------------------------------------------------------------------------------

### CODE REVIEW

**Comment by reuben** [comment]
I'm not sure we want to merge the alternative model functions since they're just bound to break when we change the code and not get any errors.

**Comment by kdavis-mozilla** [comment]
Yes we want the results, timing WER.... , but not to merge these changes.

**Comment by kdavis-mozilla** [comment]
@andi4191 @reuben I think if we want to keep these changes, we should just keep them in a branch. Opinions?

**Comment by andi4191** [comment]
Yeah I think it will be better to keep them in a separate named branch. Something like test_UiRNN instead of merging them into the master branch.

**Comment by kdavis-mozilla** [comment]
@andi4191 @reuben OK, then I'll close this and you can re-name the branch appropriately

**Comment by andi4191** [comment]
Renamed the branch to test_UiRNN

**Comment by kdavis-mozilla** [comment]
When comparing _X_ against _Y_ you run both for 10 epochs.

How do you know that _X_ and _Y_ do not converge at different rates? In other words _X_ may take 1000 epochs to converge. _Y_ may take 100. But _X_ may have a lower WER when it converges. Thus at epoch 10 _X_ may look much worse than _Y_ as it's no where near convergence. However, _X_ may in fact be much better than _Y_ when it converges.

How do you know you have the optimal hyperparameters for both _X_ and _Y_? Did you check them?
